File: web-technologies/worksheet7/stack.js
class Stack {
  constructor(size) {
    this.size = size;
    this.items = new Array(size);
    this.top = -1;
  }

  isEmpty() {
    return this.top === -1;
  }

  isFull() {
    return this.top === this.size - 1;
  }

  push(element) {
    if (this.isFull()) {
      throw new Error("Stack Overflow");
    }
    this.items[++this.top] = element;
  }

  pop() {
    if (this.isEmpty()) {
      throw new Error("Stack Underflow");
    }
    return this.items[this.top--];
  }

  peek() {
    if (this.isEmpty()) {
      throw new Error("Stack is empty");
    }
    return this.items[this.top];
  }

  find(element) {
    let tempStack = new Stack(this.size);
    let foundAt = -1;

    while (!this.isEmpty()) {
      let current = this.pop();
      if (current === element && foundAt === -1) {
        foundAt = this.top + 1;
      }
    }

    // Restore the original stack
    while (!tempStack.isEmpty()) {
      this.push(tempStack.pop());
    }

    return {
      found: foundAt !== -1,
      value: foundAt !== -1 ? element : null,
      position: foundAt,
    };
  }

  length() {
    return this.top + 1;
  }
}
export default Stack;

File: data-structures/exam-prep/stack.cpp
#include <iostream>
#include <stack>
using namespace std;

int main()
{
    stack<int> st;

    st.push(1);
    st.push(2);
    st.push(3);

    cout << st.top() << endl;

    st.pop();

    cout << st.top() << endl;

    st.pop();

    cout << st.top() << endl;

    cout << st.size() << endl;
    st.pop();

    if (st.empty()) {
        cout << "Empty" << endl;
    } else {
        cout << st.top() << endl;
    }

    cout << st.size() << endl;

    

    return 0;
}